lib/experimental/class-wp-html-processor.php

```diff
@@ -0,0 +1,720 @@
+			$insert_at,
+			' ' . self::serialize_attribute( $name, $value )
+		);
+
+		return $this;
+	}
+
+
+	public function add_class( $class_name ) {
+		$this->modifications->class_changes[ $class_name ] = true;
+	}
+
+
+	public function set_attribute( $name, $value ) {
+		if ( ! array_key_exists( WP_Tag_Find_Descriptor::comparable( $name ), $this->scanner->attributes ) ) {
+			return $this->add_attribute( $name, $value );
+		}
+
+		/** @var WP_HTML_Attribute_Token $existing */
+		$existing = $this->scanner->attributes[ $name ];
+
+		$this->modifications->replacements[] = array(
+			$existing->attribute_start,
+			$existing->attribute_end,
+			self::serialize_attribute( $name, $value )
+		);
+	}
+
+
+	public function remove_class( $class_name ) {
+		$this->modifications->class_changes[ $class_name ] = false;
+	}
+
+
+	public function remove_attribute( $name ) {
+		if ( ! array_key_exists( $name, $this->scanner->attributes ) ) {
+			return $this;
+		}
+
+		/** @var WP_HTML_Attribute_Token $existing */
+		$existing = $this->scanner->attributes[ $name ];
+
+		$this->modifications->replacements[] = array(
+			$existing->attribute_start,
+			$existing->attribute_end,
+			''
+		);
+	}
+
+
+	public static function serialize_attribute( $name, $value ) {
+		$value = str_replace( '"', "&quot;", $value );
+
+		return "{$name}=\"{$value}\"";
+	}
+
+
+	public function commit_class_changes() {
+		if ( empty( $this->modifications->class_changes ) ) {
+			return;
+		}
+
+		$existing_class = array_key_exists( 'class', $this->scanner->attributes )
+			? substr(
+				$this->scanner->document,
+				$this->scanner->attributes['class']->value->start,
+				$this->scanner->attributes['class']->value->length
+			)
+			: '';
+
+		// Remove unwanted classes.
+		$new_class = preg_replace_callback(
+			'~(?:^|[ \t])([^ \t]+)~miu',
+			function ( $matches ) {
+				list( $full_match, $class_name ) = $matches;
+
+				$comparable_name = WP_Tag_Find_Descriptor::comparable( $class_name );
+				if (
+					 array_key_exists( $comparable_name, $this->modifications->class_changes ) &&
+					 false === $this->modifications->class_changes[ $comparable_name ]
+				) {
+					 return '';
+				}
+
+				return $full_match;
+			},
+			$existing_class
+		);
+
+		// Add new classes.
+		foreach ( $this->modifications->class_changes as $class => $should_include ) {
+			if ( $should_include ) {
+				$new_class .= " {$class}";
+			}
+		}
+
+		$this->set_attribute( 'class', trim( $new_class ) );
+		$this->modifications->class_changes = array();
+	}
+
+
+	public function apply() {
+		$document = $this->scanner->document;
+
+		$this->commit_class_changes();
+
+		/*
+		 * Since we could be asking to modify attributes in a different
+		 * order than where they appear in the HTMl stream we need to
+		 * sort the modification replacements so we can progress
+		 * linearly and in a single pass.
+		 */
+		usort( $this->modifications->replacements, static function ( $a, $b ) {
+			if ( $a[0] === $b[0] ) {
```

Comment:
I noticed I keep getting lost when looking at these indices. How about turning these arrays into an instance of `WP_HTML_Replacement` or even an ad-hoc `stdclass` as follows?

```php
$a = $b = (object)array(
	'from' => $existing->attribute_start,
	'to' => $existing->attribute_end,
	'new_text' => self::serialize_attribute( $name, $value )
);
if ( $a->from === $b->from ) {
   // ... 
}
```

Reply:
turns out using an actual class for these is dramatically more memory-efficient so all other things aside I updated them in e54e75e35f5b80bd062b04223b955b2d40862416